### PR TITLE
sys-fs/bcachefs-tools: add 24_p20221120, 9999

### PR DESCRIPTION
**Pull request by Kangie**
Signed-off-by: Matt Jolly <Matt.Jolly@footclan.ninja>

### CODE REVIEW

**Comment by thesamesam** [comment]
cc @flow in case interested

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+	S="${WORKDIR}/${PN}-${MY_PV}"
+	KEYWORDS="~amd64"
+fi
+
+SLOT="0"
+LICENSE="Apache-2.0 BSD GPL-2 MIT"
+IUSE="fuse test"
+RESTRICT="!test? ( test )"
+
+DEPEND="
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:143]: No RDEPEND? is it really statically linking to all of these?

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+	SRC_URI="$(cargo_crate_uris ${CRATES})"
+else
+	MY_PV=f1f88825c371f84edb85a156de5e1962503d23b2
+	SRC_URI="https://evilpiepirate.org/git/bcachefs-tools.git/snapshot/bcachefs-tools-${MY_PV}.tar.zst
+		$(cargo_crate_uris ${CRATES})"
+	S="${WORKDIR}/${PN}-${MY_PV}"
+	KEYWORDS="~amd64"
+fi
+
+SLOT="0"
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:138]: Swap SLOT & LICENSE order

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+	winapi-x86_64-pc-windows-gnu-0.4.0
+"
+
+PYTHON_COMPAT=( python3_{8..11} )
+
+inherit cargo flag-o-matic multiprocessing python-any-r1 toolchain-funcs unpacker
+
+DESCRIPTION="Tools for bcachefs"
+HOMEPAGE="https://bcachefs.org/"
+if [[ "${PV}" = "9999" ]]; then
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:126]: 1. I prefer not quoting but it's not _wrong_ to, I prefer to just quote things which could contain a space.
2. We may as well use == and not = as we're in Bash.

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+	inherit git-r3
+	EGIT_REPO_URI="https://evilpiepirate.org/git/bcachefs-tools.git"
+	SRC_URI="$(cargo_crate_uris ${CRATES})"
+else
+	MY_PV=f1f88825c371f84edb85a156de5e1962503d23b2
+	SRC_URI="https://evilpiepirate.org/git/bcachefs-tools.git/snapshot/bcachefs-tools-${MY_PV}.tar.zst
+		$(cargo_crate_uris ${CRATES})"
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:133]: Up to you but you could unconditionally SRC_URI+= the crates line outside of the if/9999 block, but that might look weirder tbh.

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+inherit cargo flag-o-matic multiprocessing python-any-r1 toolchain-funcs unpacker
+
+DESCRIPTION="Tools for bcachefs"
+HOMEPAGE="https://bcachefs.org/"
+if [[ "${PV}" = "9999" ]]; then
+	inherit git-r3
+	EGIT_REPO_URI="https://evilpiepirate.org/git/bcachefs-tools.git"
+	SRC_URI="$(cargo_crate_uris ${CRATES})"
+else
+	MY_PV=f1f88825c371f84edb85a156de5e1962503d23b2
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:131]: I would consider MY_COMMIT instead.

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+	dev-libs/libsodium
+	dev-libs/userspace-rcu
+	sys-apps/keyutils
+	sys-apps/util-linux
+	sys-libs/zlib
+	virtual/udev
+	fuse? ( >=sys-fs/fuse-3.7.0 )
+"
+
+BDEPEND="
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:156]: Missing the eclass var for unpacker deps

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+	${PYTHON_DEPS}
+	$(python_gen_any_dep '
+		dev-python/docutils[${PYTHON_USEDEP}]
+		test? (
+			dev-python/pytest[${PYTHON_USEDEP}]
+			dev-python/pytest-xdist[${PYTHON_USEDEP}]
+		)
+	')
+	app-arch/zstd
+	sys-devel/clang
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:166]: What is Clang for?

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+	if use test; then
+		python_has_version \
+			"dev-python/pytest[${PYTHON_USEDEP}]" \
+			"dev-python/pytest-xdist[${PYTHON_USEDEP}]"
+	fi
+	python_has_version "dev-python/docutils[${PYTHON_USEDEP}]"
+
+}
+
+pkg_setup() {
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:180]: Drop?

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+	sed \
+		-e '/^CFLAGS/s:-O2::' \
+		-e '/^CFLAGS/s:-g::' \
+		-e 's:pytest-3:/bin/true:g' \
+		-i Makefile || die
+	append-lfs-flags
+}
+
+src_compile() {
+	if use fuse; then
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:201]: Optionally:
`use fuse && export BCACHEFS_FUSE=1`

or
`export BCACHEFS_FUSE=$(usex fuse)` if it actually accepts 0 as off

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+		export BCACHEFS_FUSE=1
+	fi
+	export VERSION=${PV}
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:204]:
```suggestion

	export VERSION=${PV}
```

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+	default
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:205]:
```suggestion

	default
	
```

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+	# Rust UUID-based mounter isn't in 'all' target, may as well use ebuild functions
+	for module in bch_bindgen mount; do
+		pushd "${S}/rust-src/$module"
+			LIBBCACHEFS_LIB="${S}" LIBBCACHEFS_INCLUDE="${S}" cargo_src_compile
+		popd
+	done
+	ln -f "${S}/rust-src/mount/target/release/bcachefs-mount" "${S}/mount.bcachefs" || die
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:212]:
```suggestion

	ln -f "${S}/rust-src/mount/target/release/bcachefs-mount" "${S}/mount.bcachefs" || die

```

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+	if use fuse; then
+		export BCACHEFS_FUSE=1
+	fi
+	export VERSION=${PV}
+	default
+	# Rust UUID-based mounter isn't in 'all' target, may as well use ebuild functions
+	for module in bch_bindgen mount; do
+		pushd "${S}/rust-src/$module"
+			LIBBCACHEFS_LIB="${S}" LIBBCACHEFS_INCLUDE="${S}" cargo_src_compile
+		popd
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:210]: || die

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+
+src_compile() {
+	if use fuse; then
+		export BCACHEFS_FUSE=1
+	fi
+	export VERSION=${PV}
+	default
+	# Rust UUID-based mounter isn't in 'all' target, may as well use ebuild functions
+	for module in bch_bindgen mount; do
+		pushd "${S}/rust-src/$module"
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:208]: || die and also ${module}

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+src_compile() {
+	if use fuse; then
+		export BCACHEFS_FUSE=1
+	fi
+	export VERSION=${PV}
+	default
+	# Rust UUID-based mounter isn't in 'all' target, may as well use ebuild functions
+	for module in bch_bindgen mount; do
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:207]:
```suggestion
	local module
	for module in bch_bindgen mount; do
```

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+	fi
+	EPYTEST_DESELECT=(
+		# Valgrind
+		'tests/test_fixture.py::test_read_after_free'
+		'tests/test_fixture.py::test_undefined'
+		'tests/test_fixture.py::test_write_after_free'
+		'tests/test_fixture.py::test_undefined_branch'
+		'tests/test_fixture.py::test_leak'
+		'tests/test_fixture.py::test_check'
+		# Fails in portage
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:230]: Elaborate on why, maybe?

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+		'tests/test_basic.py::test_fsck'
+		'tests/test_basic.py::test_list'
+		'tests/test_basic.py::test_list_inodes'
+		'tests/test_basic.py::test_list_dirent'
+	)
+	epytest -v -n "$(makeopts_jobs)"
+}
+
+src_install() {
+	QA_FLAGS_IGNORED="/usr/bin/mount.bcachefs"
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:241]: Needs to be relative, not absolute. Also, normally defined in global scope.

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+	# Raised upstream; we don't expect anything to link against this outside of bcachefs-tools bins, for now
+	QA_SONAME=".*libbcachefs.so"
+	exeinto /usr/bin
+	for file in bcachefs fsck.bcachefs mkfs.bcachefs mount.bcachefs mount.bcachefs.sh; do
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:245]: Always `local`ize vars so they don't leak:
```suggestion
	local
	for file in bcachefs fsck.bcachefs mkfs.bcachefs mount.bcachefs mount.bcachefs.sh; do
```

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+		doexe $file
+	done
+	dolib.so libbcachefs.so
+	doman bcachefs.8
+}
+
+pkg_postinst() {
+	if use fuse; then
+		ewarn
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:254]:
```suggestion
```

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 20]
```diff
@@ -0,0 +1,259 @@
...
+		ewarn "FUSE support is experimental."
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:255]: I would then package.use.mask the flag too, tbh.

**Review by thesamesam — CHANGES_REQUESTED** [review]

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+BDEPEND="
+	${PYTHON_DEPS}
+	$(python_gen_any_dep '
+		dev-python/docutils[${PYTHON_USEDEP}]
+		test? (
+			dev-python/pytest[${PYTHON_USEDEP}]
+			dev-python/pytest-xdist[${PYTHON_USEDEP}]
+		)
+	')
+	app-arch/zstd
```
Comment (thesamesam) [inline sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:165]: (I expect we'll be dropping this and using the unpacker var)

File: sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild (hunk 20 above)
Comment (Kangie) [reply sys-fs/bcachefs-tools/bcachefs-tools-24_p20221120.ebuild:255]: It's less 'it's definitely going to eat your data' and more 'do this at your own risk' as I understand it. I'm inclined to leave it unmasked for those that *want* to use it, but if you feel strongly I'm equally happy to mask

**Comment by Flowdalic** [comment]
> cc @Flowdalic in case interested
after bcachefs got merged into the mainline kernel :)

**Pull request closed without merging**